Source/WebCore/Modules/mediastream/gstreamer/GStreamerDtlsTransportBackend.cpp

```diff
@@ -29,8 +29,80 @@
 
 namespace WebCore {
 
-GStreamerDtlsTransportBackend::GStreamerDtlsTransportBackend(const GRefPtr<GstWebRTCDTLSTransport>& transport)
-    : m_backend(transport)
+class GStreamerDtlsTransportBackendObserver final : public ThreadSafeRefCounted<GStreamerDtlsTransportBackendObserver> {
+public:
+    static Ref<GStreamerDtlsTransportBackendObserver> create(RTCDtlsTransportBackend::Client& client, GRefPtr<GstWebRTCDTLSTransport>&& backend) { return adoptRef(*new GStreamerDtlsTransportBackendObserver(client, WTFMove(backend))); }
+
+    void start();
+    void stop();
+
+private:
+    GStreamerDtlsTransportBackendObserver(RTCDtlsTransportBackend::Client&, GRefPtr<GstWebRTCDTLSTransport>&&);
+
+    void stateChanged();
+
+    GRefPtr<GstWebRTCDTLSTransport> m_backend;
+    WeakPtr<RTCDtlsTransportBackend::Client> m_client;
+};
+
+GStreamerDtlsTransportBackendObserver::GStreamerDtlsTransportBackendObserver(RTCDtlsTransportBackend::Client& client, GRefPtr<GstWebRTCDTLSTransport>&& backend)
+    : m_backend(WTFMove(backend))
+    , m_client(client)
+{
+    ASSERT(m_backend);
+}
+
+void GStreamerDtlsTransportBackendObserver::stateChanged()
+{
+    if (!m_client)
+        return;
+
+    callOnMainThread([this, protectedThis = Ref { *this }]() mutable {
+        if (!m_client || !m_backend)
+            return;
+
+        GstWebRTCDTLSTransportState state;
+        g_object_get(m_backend.get(), "state", &state, nullptr);
+
+        Vector<Ref<JSC::ArrayBuffer>> certificates;
+
+        // Access to DTLS certificates is not memory-safe in GStreamer versions older than 1.22.3.
+        // See also: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/commit/d9c853f165288071b63af9a56b6d76e358fbdcc2
+        if (webkitGstCheckVersion(1, 22, 3)) {
+            GUniqueOutPtr<char> remoteCertificate;
+            GUniqueOutPtr<char> certificate;
+            g_object_get(m_backend.get(), "remote-certificate", &remoteCertificate.outPtr(), "certificate", &certificate.outPtr(), nullptr);
+
+            certificates.reserveInitialCapacity(2);
+            if (remoteCertificate) {
+                auto remoteCertificateString = makeString(remoteCertificate.get());
+                auto jsRemoteCertificate = JSC::ArrayBuffer::create(remoteCertificateString.characters8(), remoteCertificateString.sizeInBytes());
+                certificates.uncheckedAppend(WTFMove(jsRemoteCertificate));
+
+                auto certificateString = makeString(certificate.get());
```

Comment:
I guess it's guaranteed that if the first exists, the second as well, otherwise you would need to change this.

Reply:
Unclear actually. I'll add an `if`.